### PR TITLE
Add Average Precision to metrics

### LINKED ISSUE #8085: Add Average Precision and R2 to the metrics
### Discussed in https://github.com/Project-MONAI/MONAI/discussions/8081

<div type='discussions-op-text'>

<sup>Originally posted by **thibaultdvx** September 13, 2024</sup>
Hi, 
I would like to suggest to add the [**Average Precision**](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html) in the classification metrics. I think that the implementation would be very similar to ROC AUC, so it wouldn't be a big deal to integrate it.
I would be happy to make a PR to do so.

I also feel that an important regression metric is missing:[ **R2 score**](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html). Is there any reason why?
Once again, I would be happy to help to implement it!

WDYT?</div>

### PR DESCRIPTION
**Pull request by thibaultdvx**
Fixes #8085.

### Description

Average Precision is very similar to ROCAUC, so I was very much inspired by the ROCAUC implementation.
More precisely, I created:
- `AveragePrecisionMetric` and `compute_average_precision` in `monai.metrics`,
- a handler called `AveragePrecision` in `monai.handlers`,
- three unittest modules: `test_compute_average_precision.py`, `test_handler_average_precision.py` and `test_handler_average_precision_dist.py`.

I also modified the docs to mention Average Precision.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by thibaultdvx** [comment]
hi @KumoLiu, could you please have a look at this PR?

**Review by KumoLiu — COMMENTED** [review]
Thanks for the pr, overall looks good to me.

File: monai/metrics/average_precision.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+import torch
+
+from monai.utils import Average, look_up_option
+
+from .metric import CumulativeIterationMetric
+
+
+class AveragePrecisionMetric(CumulativeIterationMetric):
+    """
+    Computes Average Precision (AP). Referring to: `sklearn.metrics.average_precision_score
```
Comment (ericspod) [inline monai/metrics/average_precision.py:31]: For the lazy it would be nice to explain here what this metric is and its intended application area.

**Review by ericspod — APPROVED** [review]
Looks good to me too, I had a minor comment and it would be nice to combine the last 2 test files, it would be better to have fewer test files if possible.

**Comment by ericspod** [comment]
Hi @thibaultdvx we've updated our tests with a directory structure, if you could please look over comments and update where your test files are we can come back to this PR. Thanks!

**Comment by thibaultdvx** [comment]
Hi @ericspod @KumoLiu, sorry for the absence. I got work and forgot this PR. I'll get right on it!

**Comment by ericspod** [comment]
Hi @thibaultdvx I'm going to merge recent changes into your branch and see if that sorts the test issues, I don't think it was related to PR either.

File: tests/metrics/test_compute_average_precision.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            TEST_CASE_5,
+            TEST_CASE_6,
+            TEST_CASE_7,
+            TEST_CASE_8,
+            TEST_CASE_9,
+            TEST_CASE_10,
+            TEST_CASE_11,
+            TEST_CASE_12,
+        ]
+    )
```
Comment (ericspod) [inline tests/metrics/test_compute_average_precision.py:161]:
```suggestion
]

ALL_TESTS = [
    TEST_CASE_1,
    TEST_CASE_2,
    TEST_CASE_3,
    TEST_CASE_4,
    TEST_CASE_5,
    TEST_CASE_6,
    TEST_CASE_7,
    TEST_CASE_8,
    TEST_CASE_9,
    TEST_CASE_10,
    TEST_CASE_11,
    TEST_CASE_12,
]

class TestComputeAveragePrecision(unittest.TestCase):

    @parameterized.expand(ALL_TESTS)
    def test_value(self, y_pred, y, softmax, to_onehot, average, expected_value):
        y_pred_trans = Compose([ToTensor(), Activations(softmax=softmax)])
        y_trans = Compose([ToTensor(), AsDiscrete(to_onehot=to_onehot)])
        y_pred = torch.stack([y_pred_trans(i) for i in decollate_batch(y_pred)], dim=0)
        y = torch.stack([y_trans(i) for i in decollate_batch(y)], dim=0)
        result = compute_average_precision(y_pred=y_pred, y=y, average=average)
        np.testing.assert_allclose(expected_value, result, rtol=1e-5)

    @parameterized.expand(ALL_TESTS)
```

**Comment by ericspod** [comment]
I think this is good to go now, @KumoLiu if you could please trigger blossom, thanks.

**Comment by KumoLiu** [comment]
Hi @thibaultdvx, please help fix the format issue: https://github.com/Project-MONAI/MONAI/actions/runs/13406057032/job/37446075381?pr=8089
Then I will help trigger the test again, thanks!
